### PR TITLE
customcommands: draft API proposal / specification

### PR DESCRIPTION
**Pull request by l-zeuch**
This draft pull request proposes an API to manage custom commands.

It serves as a place to discuss thoughts, ideas, and questions from the community.

For ease of viewing the markdown file, please follow [this link](https://github.com/l-zeuch/yagpdb/blob/feat/cc-api/customcommands/cc-api.md).

If you have direct feedback, feel free to use the "review" feature and leave a comment (or, if you're a maintainer, request changes as you deem necessary).

For general questions, just leave a general comment. For *just* voting, please use reactions, as the "+1"/"-1" comments tend to drown actual conversations, as they are rather spammy.

### CODE REVIEW

File: customcommands/api.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+// serving as a wrapper around the same.
+type APICustomCommand struct {
+	ID            int64              `json:"id"`
+	Name          string             `json:"name" valid:",0,100"`
+	IsEnabled     bool               `json:"is_enabled"`
+	TriggerType   CommandTriggerType `json:"trigger_type"`
+	Trigger       string             `json:"trigger" valid:",0,1000"`
+	CaseSensitive bool               `json:"case_sensitive"`
+	Responses     []string           `json:"responses" valid:"template,10000"`
+}
```
Comment (jo3-l) [inline customcommands/api.go:61]: `valid` tags aren't necessary here.

File: customcommands/api.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		// Doesn't have premium, send 403.
+		if !ok {
+			http.Error(w, "Guild does not have premium.", http.StatusForbidden)
+			return
+		}
```
Comment (jo3-l) [inline customcommands/api.go:86]:
```suggestion
		isPrem, err := premium.IsGuildPremium(activeGuild.ID)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}

		if !isPrem {
			http.Error(w, "Guild does not have premium.", http.StatusForbidden)
			return
		}
```
Renaming `ok` to something a little more descriptive obviates the need for the comment.

**Review by jo3-l — COMMENTED** [review]
I'm aware this isn't fully done yet — just wanted to leave these comments before I forget :-) Feel free to address whenever ready.

File: customcommands/api.go (hunk 1 above)
Comment (l-zeuch) [reply customcommands/api.go:61]: Honestly, I'm not even sure whether this whole struct is needed. There's a few questions I have to solve, the main one is how to serialise the `models.CustomCommand` struct into the above.

I'll remove the tags in the coming commit(s) and see how to proceed from there.

**Comment by l-zeuch** [comment]
> I'm aware this isn't fully done yet — just wanted to leave these comments before I forget :-) Feel free to address whenever ready.

Thanks for the feedback, I appreciate it! I prefer small commits over a huge chunk either way, that way it is easier for all involved to review and address feedback.

**Comment by l-zeuch** [comment]
I'll be closing this PR for the time being as to not fill the list with
junk. I'll reopen it when I have the time to work on it again and
refined my Go skills a bit more — this is not to say it won't happen,
but as explained in my initial comment we want to get it right the first
time and not rush it.

The proposal itself could also use some refining, so I'll be working on
that with a higher priority than implementing the API itself — there's
still some things I am yet to learn about the framework(s) we use, so I
am going to take my time with this.

If you have any resources on hand that could help me with this, please
do share them with me.

**Pull request closed without merging**